[NEW] Save unfinished message

Pancor opened the pull request:
@RocketChat/android

Closes #933 

From now, unfinished messages are saved and ready to show up again when user reopen chat room. I added `showSoftKeyboardOnActivityStart()` function to `KeyboardHelper`, becouse the given one (`showSoftKeyboard()`) doesn't want to open soft keyboard at start of a activity/fragment.

![example](https://user-images.githubusercontent.com/24636767/37845188-364e52f0-2eca-11e8-9022-0c499cdc0798.gif)

Pancor committed:
adds const

Pancor committed:
add save/get functions

Pancor committed:
add show soft keyboard at activity start functionality

Pancor committed:
refactor code and add clear message possibility

Pancor committed:
add https

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat.Android?pullRequest=958) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat.Android?pullRequest=958) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: leonardoaramaki<br/>:white_check_mark: filipedelimabrito<br/>:x: Pancor<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat.Android?pullRequest=958) it.</sub>

Pancor committed:
Merge branch 'develop-2.x' into feature/save-unfinished-message

Pancor committed:
add missing bracket

leonardoaramaki commented on app/src/main/java/chat/rocket/android/helper/KeyboardHelper.kt:
```diff
@@ -50,4 +51,15 @@ object KeyboardHelper {
+     *
+     *  @param view View to receive input focus.
+     */
+    fun showSoftKeyboardOnActivityStart(activity: Activity, view: View) {
```
The soft keyboard is not showing probably because of this line: https://github.com/RocketChat/Rocket.Chat.Android/blob/45ea38242a208389104c6d9e2baa026df76e71be/app/src/main/AndroidManifest.xml#L55.
Maybe if you remove the `stateAlwaysHidden` flag it will show the keyboard. I think it's better than this approach. :+1:

leonardoaramaki commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -49,7 +49,8 @@ class ChatRoomPresenter @Inject constructor(private val view: ChatRoomView,
                                             factory: ConnectionManagerFactory,
-                                            private val mapper: ViewModelMapper) {
+                                            private val mapper: ViewModelMapper,
+                                            private val localRepository: LocalRepository) {
```
Can you remove the duplicated `localRepository` here?

leonardoaramaki commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -553,4 +554,33 @@ class ChatRoomPresenter @Inject constructor(private val view: ChatRoomView,
+     * @param unfinishedMessage The unfinished message to save.
+     */
+    fun saveUnfinishedMessage(chatRoomId: String, unfinishedMessage: String) {
+        val key = LocalRepository.UNFINISHED_MSG_KEY + chatRoomId
```
We're going to support multiserver on the next release. So here the key should be:
`val key = "${currentServer}_${LocalRepository.UNFINISHED_MSG_KEY}$chatRoomId"`

leonardoaramaki commented on app/src/main/java/chat/rocket/android/chatroom/presentation/ChatRoomPresenter.kt:
```diff
@@ -553,4 +554,33 @@ class ChatRoomPresenter @Inject constructor(private val view: ChatRoomView,
+     * @return Returns the unfinished message.
+     */
+    fun getUnfinishedMessage(chatRoomId: String): String {
+        val key = LocalRepository.UNFINISHED_MSG_KEY + chatRoomId
```
The same here:
`val key = "${currentServer}_${LocalRepository.UNFINISHED_MSG_KEY}$chatRoomId"`

leonardoaramaki committed:
Merge branch 'develop-2.x' into feature/save-unfinished-message

Pancor commented:
Well, after removing `stateAlwaysHidden` from `Manifest.xml` file and replacing my added function `showSoftKeyboardOnActivityStart()` to `showSoftKeyboard()` the soft keyboard shows and immediately closes.

![keyboard](https://user-images.githubusercontent.com/24636767/37926878-d04e2e46-3138-11e8-9be4-bb88ea4eadb7.gif)

But if I add some delay, for example:
`handler.postDelayed({
                KeyboardHelper.showSoftKeyboard(text_message)
            }, 500)`,
it works, but I think, that's not the best aproach of resolving the problem. If I use my function `showSoftKeyboardOnActivityStart()`, it works. 
Tomorrow I can investigate, why the soft keybord is hidding, when using `showSoftKeyboard()` from `KeyboardHelper`.

leonardoaramaki commented:
@Pancor are you still working on this?

Pancor committed:
add support for muliserver

Pancor committed:
code refactor

Pancor committed:
remove unused function

Pancor committed:
remove stateAlwaysHidden

Pancor committed:
Merge branch 'develop-2.x' into feature/save-unfinished-message

Pancor commented:
Yup, I'm still working on this, It takes me some time, becouse of the one problem, that I've got. I added support for multiserver , removed duplicated `localRepository` and refactored `ChatRoomFragment` to use another function from `KeyboardHelper`.
The problem, that I have is about showing saved message in landscape orientation, when keyboard is in fullscreen mode. I can't find any method, that could help me to show keyboard in landscape orientation.

leonardoaramaki commented:
@Pancor In landscape orientation one thing we need to do is to set to keyboard not to be fullscreen. I'm going to work on that soon. For the time being can you solve the conflicts?

Pancor committed:
Merge branch 'develop-2.x' into feature/save-unfinished-message

Pancor commented:
@leonardoaramaki  conflicts colved

leonardoaramaki committed:
Merge branch 'develop-2.x' into feature/save-unfinished-message

rafaelks commented:
@Pancor Still have some conflicts. Is this ready to be reviewed? Please fix the conflicts :-)

Pancor committed:
Merge branch 'develop-2.x' into feature/save-unfinished-message

leonardoaramaki committed:
Merge branch 'feature/save-unfinished-message' of https://github.com/Pancor/Rocket.Chat.Android into Pancor-feature/save-unfinished-message

leonardoaramaki committed:
Merge branch 'develop' into feature/save-unfinished-message

leonardoaramaki committed:
Merge branch 'develop' into feature/save-unfinished-message

philipbrito committed:
Merge branch 'develop' into feature/save-unfinished-message

leonardoaramaki committed:
Merge branch 'develop' into feature/save-unfinished-message
